<h1>Table of Contents<span class="tocSkip"></span></h1>

<div class="toc"><ul class="toc-item">
    <li><span>
    <a href="#Initial_Data_Analysis" data-toc-modified-id="Initial_Data_Analysis-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Initial_Data_Analysis</a>
    </span></li>
    <li><span>
    <a href="#Load,-rename-columns,-information" data-toc-modified-id="Load,-rename-columns,-information-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Load, rename columns, information</a>
    </span></li>
    <li><span>
    <a href="#Check-number-of-null-values" data-toc-modified-id="Check-number-of-null-values-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Check number of null values</a>
    </span>
        <ul class="toc-item">
            <li><span><a href="#NOTE-1:" data-toc-modified-id="NOTE-1:-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>NOTE 1:</a></span></li>
            <li><span><a href="#NOTE-2:" data-toc-modified-id="NOTE-2:-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>NOTE 2:</a></span></li>
            <li><span><a href="#NOTE-3:" data-toc-modified-id="NOTE-3:-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>NOTE 3:</a></span></li>
        </ul></li>
    <li><span>
    <a href="#Imputation-using-column-means" data-toc-modified-id="Imputation-using-column-means-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Imputation using column means</a>
    </span>
    </li>
    <li><span>
    <a href="#Re-Check-null-data-points" data-toc-modified-id="Re-Check-null-data-points-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Re-Check null data points</a>
    </span>
    </li>
    <li><span>
    <a href="#Save-intermediate-dataframe" data-toc-modified-id="Save-intermediate-dataframe-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Save intermediate dataframe</a>
    </span>
    </li>
    </ul>
</div>

# Life_Expectancy_WHO_UN_Analysis_Modeling

## Initial_Data_Analysis

To:&nbsp;&nbsp;&nbsp;&nbsp; [Magnimind](https://magnimindacademy.com/)

From: Matt Curcio, matt.curcio.ri@gmail.com

Date: 2023-01-29

Re:&nbsp;&nbsp;&nbsp; NOTEBOOK #2

---

Purpose: This notebook investigates *missing values, renames features, and imputes missing values by column mean*.

Input: `Life_Expectancy_Data.csv`

Output: `Clean_LE_Data_w_Means_1.csv`

In [16]:
# Common Python Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

## Load, rename columns, dataset

In [17]:
filename = '/home/mcc/Desktop/Life_Expectancy_Analysis_Modeling/data/raw/Life_Expectancy_Data.csv'

column_names = ['Country','Year','Status','LifeExpectancy','AdultMort',
                'InfD','EtOH','PercExpen','HepB','Measles',
                'BMI','lt5yD','Polio','TotalExpen','DTP','HIV',
                'GDP','Population','Thin1_19y','Thin5_9y','Income',
                'Education']

df = pd.read_csv(filename, names=column_names, header=0)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         2938 non-null   object 
 1   Year            2938 non-null   int64  
 2   Status          2938 non-null   object 
 3   LifeExpectancy  2928 non-null   float64
 4   AdultMort       2928 non-null   float64
 5   InfD            2938 non-null   int64  
 6   EtOH            2744 non-null   float64
 7   PercExpen       2938 non-null   float64
 8   HepB            2385 non-null   float64
 9   Measles         2938 non-null   int64  
 10  BMI             2904 non-null   float64
 11  lt5yD           2938 non-null   int64  
 12  Polio           2919 non-null   float64
 13  TotalExpen      2712 non-null   float64
 14  DTP             2919 non-null   float64
 15  HIV             2938 non-null   float64
 16  GDP             2490 non-null   float64
 17  Population      2286 non-null   f

## Check number of null values

In [18]:
print('='*38)
print('Column              No. missing values')
print('='*38)
print(df.isnull().sum())

Column              No. missing values
Country             0
Year                0
Status              0
LifeExpectancy     10
AdultMort          10
InfD                0
EtOH              194
PercExpen           0
HepB              553
Measles             0
BMI                34
lt5yD               0
Polio              19
TotalExpen        226
DTP                19
HIV                 0
GDP               448
Population        652
Thin1_19y          34
Thin5_9y           34
Income            167
Education         163
dtype: int64


### NOTE 1: 
- The **Dependent feature** 'LifeExpectancy' has 10 missing values. 
 
- Q. What exactly is missing from LifeExpectancy column?

In [19]:
df.loc[df['LifeExpectancy'].isnull()]

Country  Year      Status  LifeExpectancy  AdultMort  \
624            Cook Islands  2013  Developing             NaN        NaN   
769                Dominica  2013  Developing             NaN        NaN   
1650       Marshall Islands  2013  Developing             NaN        NaN   
1715                 Monaco  2013  Developing             NaN        NaN   
1812                  Nauru  2013  Developing             NaN        NaN   
1909                   Niue  2013  Developing             NaN        NaN   
1958                  Palau  2013  Developing             NaN        NaN   
2167  Saint Kitts and Nevis  2013  Developing             NaN        NaN   
2216             San Marino  2013  Developing             NaN        NaN   
2713                 Tuvalu  2013  Developing             NaN        NaN   

      InfD  EtOH   PercExpen  HepB  Measles  ...  Polio  TotalExpen   DTP  \
624      0  0.01    0.000000  98.0        0  ...   98.0        3.58  98.0   
769      0  0.01   11.419555  96.0        0  ...   96.0        5.58  96.0   
1650     0  0.01  871.878317   8.0        0  ...   79.0       17.24  79.0   
1715     0  0.01    0.000000  99.0        0  ...   99.0        4.30  99.0   
1812     0  0.01   15.606596  87.0        0  ...   87.0        4.65  87.0   
1909     0  0.01    0.000000  99.0        0  ...   99.0        7.20  99.0   
1958     0   NaN  344.690631  99.0        0  ...   99.0        9.27  99.0   
2167     0  8.54    0.000000  97.0        0  ...   96.0        6.14  96.0   
2216     0  0.01    0.000000  69.0        0  ...   69.0        6.50  69.0   
2713     0  0.01   78.281203   9.0        0  ...    9.0       16.61   9.0   

      HIV          GDP  Population  Thin1_19y  Thin5_9y  Income  Education  
624   0.1          NaN         NaN        0.1       0.1     NaN        NaN  
769   0.1   722.756650         NaN        2.7       2.6   0.721       12.7  
1650  0.1  3617.752354         NaN        0.1       0.1     NaN        0.0  
1715  0.1          NaN         NaN        NaN       NaN     NaN        NaN  
1812  0.1   136.183210         NaN        0.1       0.1     NaN        9.6  
1909  0.1          NaN         NaN        0.1       0.1     NaN        NaN  
1958  0.1  1932.122370       292.0        0.1       0.1   0.779       14.2  
2167  0.1          NaN         NaN        3.7       3.6   0.749       13.4  
2216  0.1          NaN         NaN        NaN       NaN     NaN       15.1  
2713  0.1  3542.135890      1819.0        0.2       0.1     NaN        0.0  

[10 rows x 22 columns]

In [20]:
# Produces one entry
df.loc[df['Country'] == 'Cook Islands']

Country  Year      Status  LifeExpectancy  AdultMort  InfD  EtOH  \
624  Cook Islands  2013  Developing             NaN        NaN     0  0.01   

     PercExpen  HepB  Measles  ...  Polio  TotalExpen   DTP  HIV  GDP  \
624        0.0  98.0        0  ...   98.0        3.58  98.0  0.1  NaN   

     Population  Thin1_19y  Thin5_9y  Income  Education  
624         NaN        0.1       0.1     NaN        NaN  

[1 rows x 22 columns]

### Note 2:

- Further investigation (not shown here) finds that all ten missing values are from countries with only one row (year=2013). 

- 'LifeExpectancy' is a Dependent variable. I will delete(drop) those 10 observations with NO labels rather than impute them.

- 'LifeExpectancy' appears to be **Missing Completely at Random(MCAR)**. The advantage of **MCAR** is that the analysis is unbiased.

In [21]:
# 10 country rows deleted
df.dropna(subset=['LifeExpectancy'], inplace=True)

In [22]:
print('\nDataframe shape:', df.shape)

df.isnull().sum()


Dataframe shape: (2928, 22)


Country             0
Year                0
Status              0
LifeExpectancy      0
AdultMort           0
InfD                0
EtOH              193
PercExpen           0
HepB              553
Measles             0
BMI                32
lt5yD               0
Polio              19
TotalExpen        226
DTP                19
HIV                 0
GDP               443
Population        644
Thin1_19y          32
Thin5_9y           32
Income            160
Education         160
dtype: int64

### NOTE 2: 
- Three features with the highest number of mising vlaues (out of 2928) are:

|  | Feature | Number Missing | % Missing |
|--|:--------|---------------:|----------:|
|1 | Population | 644 | 22.0% |
|2 | HepB | 553 |  18.9% |
|3 | GDP | 448 | 15.3% |

- ['Population', 'HepB', 'GDP'] are missing greater than 15%.

- Because these three features are missing more than 15%, they will be deleted from analysis.

In [23]:
df.drop(['Population', 'HepB', 'GDP'], axis=1, inplace=True)

## Imputation using column means

In [24]:
df['InfD'].fillna(np.mean(df.InfD), inplace=True)
df['EtOH'].fillna(np.mean(df.EtOH), inplace=True)
df['PercExpen'].fillna(np.mean(df.PercExpen), inplace=True)
df['Measles'].fillna(np.mean(df.Measles), inplace=True)
df['BMI'].fillna(np.mean(df.BMI), inplace=True)
df['Polio'].fillna(np.mean(df.Polio), inplace=True)
df['TotalExpen'].fillna(np.mean(df.TotalExpen), inplace=True)
df['DTP'].fillna(np.mean(df.DTP), inplace=True)
df['Thin1_19y'].fillna(np.mean(df.Thin1_19y), inplace=True)
df['Thin5_9y'].fillna(np.mean(df.Thin5_9y), inplace=True)
df['Income'].fillna(np.mean(df.Income), inplace=True)
df['Education'].fillna(np.mean(df.Education), inplace=True)

# Convert Dev status to a binary variable, where Developing Nations = 0, Developed Nation status = 1.
df['Status'] = df['Status'].apply(lambda x: 0 if x == 'Developing' else 1).astype('int8')

## Re-Check null data points

In [25]:
print()
print('='*20,'\n')

print('\nShape of Cleaned and Imputed dataframe:', df.shape)
      
print(df.info())




Shape of Cleaned and Imputed dataframe: (2928, 19)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 2928 entries, 0 to 2937
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         2928 non-null   object 
 1   Year            2928 non-null   int64  
 2   Status          2928 non-null   int8   
 3   LifeExpectancy  2928 non-null   float64
 4   AdultMort       2928 non-null   float64
 5   InfD            2928 non-null   int64  
 6   EtOH            2928 non-null   float64
 7   PercExpen       2928 non-null   float64
 8   Measles         2928 non-null   int64  
 9   BMI             2928 non-null   float64
 10  lt5yD           2928 non-null   int64  
 11  Polio           2928 non-null   float64
 12  TotalExpen      2928 non-null   float64
 13  DTP             2928 non-null   float64
 14  HIV             2928 non-null   float64
 15  Thin1_19y       2928 non-null   float64
 16  Thin5_9y        2928 non

In [26]:
df.head()

Country  Year  Status  LifeExpectancy  AdultMort  InfD  EtOH  \
0  Afghanistan  2015       0            65.0      263.0    62  0.01   
1  Afghanistan  2014       0            59.9      271.0    64  0.01   
2  Afghanistan  2013       0            59.9      268.0    66  0.01   
3  Afghanistan  2012       0            59.5      272.0    69  0.01   
4  Afghanistan  2011       0            59.2      275.0    71  0.01   

   PercExpen  Measles   BMI  lt5yD  Polio  TotalExpen   DTP  HIV  Thin1_19y  \
0  71.279624     1154  19.1     83    6.0        8.16  65.0  0.1       17.2   
1  73.523582      492  18.6     86   58.0        8.18  62.0  0.1       17.5   
2  73.219243      430  18.1     89   62.0        8.13  64.0  0.1       17.7   
3  78.184215     2787  17.6     93   67.0        8.52  67.0  0.1       17.9   
4   7.097109     3013  17.2     97   68.0        7.87  68.0  0.1       18.2   

   Thin5_9y  Income  Education  
0      17.3   0.479       10.1  
1      17.5   0.476       10.0  
2      17.7   0.470        9.9  
3      18.0   0.463        9.8  
4      18.2   0.454        9.5

## Save intermediate dataframe

In [27]:
path = '/home/mcc/Desktop/Life_Expectancy_Analysis_Modeling/data/processed/'
fileName = "Clean_LE_Data_w_Means_2.csv"

df.to_csv(path+fileName, index=False)

print('Clean_LE_Data_w_Means_2.csv: SAVED')

Clean_LE_Data_w_Means_2.csv: SAVED
